sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DebugAssertHasNoSideEffects.cs

```diff
@@ -53,7 +53,8 @@ protected override void Initialize(SonarAnalysisContext context)
                 {
                     var invokedMethodSyntax = c.Node as InvocationExpressionSyntax;
 
-                    if (ContainsCallsWithSideEffects(invokedMethodSyntax) &&
+                    if (invokedMethodSyntax.ToStringContains("Assert") &&
```

Comment:
This also matches when word "Assert" appears in argument list, even as a part of string literal.

Let's put IsDebugAssert first. Move the check inside. The new check is anyway part of that method. And make it little more precise:
```C#
private static bool IsDebugAssert(InvocationExpressionSyntax invocation, SyntaxNodeAnalysisContext context) =>
    invocation.Expression is MemberAccessExpressionSyntax memberAccess
        && memberAccess.Name.Identifier.ValueText == nameof(System.Diagnostics.Debug.Assert)
        && context.SemanticModel.GetSymbolInfo(invocation).Symbol is IMethodSymbol symbol
        && symbol.IsDebugAssert();
```

Evaluating all DescendantNodes of all invocations first could be also expensive, as it's currently implemented.